packages/taiko-client/prover/proof_producer/combined_producer.go

```diff
@@ -17,24 +17,143 @@ import (
 	"github.com/taikoxyz/taiko-mono/packages/taiko-client/bindings/metadata"
 )
 
+type BlockProofState struct {
+	verifiedTiers []uint16
+	proofs        []encoding.SubProof
+}
+
+type ProofStateManager struct {
+	mu     sync.Mutex
+	states map[uint64]*BlockProofState
+}
+
+func NewProofStateManager() *ProofStateManager {
+	return &ProofStateManager{
+		states: make(map[uint64]*BlockProofState),
+	}
+}
+
+func (m *ProofStateManager) create(blockID *big.Int) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		state = &BlockProofState{
+			verifiedTiers: []uint16{},
+			proofs:        []encoding.SubProof{},
+		}
+		m.states[blockIDUint64] = state
+	}
+}
+
+func (m *ProofStateManager) containsTier(blockID *big.Int, tier uint16) bool {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		return false
+	}
+
+	return slices.Contains(state.verifiedTiers, tier)
+}
+
+// addTierAndProof marks the tier as verified and adds the subproof to the block's state if
+// the state has not yet collected enough proofs. It returns whether the required number
+// of proofs has now been reached.
+func (m *ProofStateManager) addTierAndProof(
+	blockID *big.Int,
+	tier uint16,
+	subProof encoding.SubProof,
+	requiredProofs uint8,
+) (reachedRequired bool) {
+	blockIDUint64 := blockID.Uint64()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[blockIDUint64]
+	if !ok {
+		// create if it doesn't exist
```

Comment:
IMO there isn't any value if this comment here as it is obvious from the code, so I'd just delete it as it adds noise.

If you want to be more explicit you could do 

```Go
state, alreadyExsists := m.states[blockIDUint64]
if !alreadyExists {}
```

But this is a bit _unidomatic_ to do in Go.